### DISCUSSION TITLE
Read from AsyncTcpSocket in separate Thread

### DISCUSSION THREAD

**Discussion by iEAmi** [activej/activej#263] [Q&A] [answered]
Hi.
How can I split read and write (each one in dedicate thread) for AsyncTcpSocket?

**Comment by eduard-vasinskyi** [comment, marked as answer]
Hi, @iEAmi

`AsyncTcpSocket` is meant to be used inside an `Eventloop`. An `Eventloop` runs in a single thread, so every operation on `AsyncTcpSocket` should be performed from within that thread. This is the main idea behind using `AsyncTcpSocket` in the first place.

However, if you absolutely need to read from the `AsyncTcpSocket` in a separate thread, you can use a converter. First, you would need to convert the socket to a `ChannelSupplier`. Then you can convert a `ChannelSupplier` to a regular `InputStream` which can be used from a non-Eventloop thread. 

A code could look something like this:
```java
Eventloop eventloop = Eventloop.getCurrentEventloop();
ChannelSupplier<ByteBuf> channelSupplier = ChannelSupplier.ofSocket(socket);
 
Thread thread = new Thread(() -> {
  try (InputStream inputStream = ChannelSuppliers.channelSupplierAsInputStream(eventloop, channelSupplier)) {
     // read from 'inputStream'
  }
});

thread.start();
```

Although, the performance will be worse than reading directly inside the Eventloop thread. I would suggest reading the data from the socket inside the Eventloop thread, and then passing all the read data to a separate thread for processing.
